In [1]:
#importing libraries
import pandas as pd
import numpy as np
import re

In [2]:
#Load the dataset into a pandas DataFrame
data = pd.read_csv('messy_data.csv')

In [3]:
#creating datframe
df = pd.DataFrame(data)

In [4]:
#Inspecting the data structure by displaying the first few rows
df.head(10)

Unnamed: 0                                    ID                     Name  \
0           0  1e407ff9-6255-489d-a0de-34135d4f74bd            Hunter Thomas   
1           1  379f55b8-87d5-4739-a146-7400b78c24d1             Jeremy Irwin   
2           2  18261368-dfa1-47f0-afc6-bddf45926b07  Jennifer Hammondquickly   
3           3  ae7cf7cf-17cf-4c8b-9c44-4f61a9a238e5          Sydney Taylorso   
4           4  14ed3e6a-e0f5-4bbe-8d93-8665267f5c90                Julia Lee   
5           5  f92ea13f-24d3-4792-91e4-5b0e07d90c2c              Lisa Bailey   
6           6  c4892dbb-2554-4ce1-8f11-f6597d7f6782       Patricia Brown DVM   
7           7  465e625b-f924-4061-8356-c7e3130f7118             Derek Murphy   
8           8  12cc2f16-d641-402e-b22b-3332bafc4f1d                      NaN   
9           9  d06f507d-62e5-4bd6-a252-4e60e109d324               Jay Fisher   

    Age                         Email   Join Date         Salary   Department  
0  25.0            xlopez@hotmail.com         NaN   88552.000000        Sales  
1  90.0               Jillian Jenkins  2022-07-07  139227.000000          NaN  
2  66.0               jscottgreen.biz  2023-11-21   65550.000000  Engineering  
3  39.0            luke56gonzalez.com  2021-11-05  139932.000000     SupportJ  
4  71.0       figueroakayla@yahoo.com         NaN  143456.000000    Marketing  
5  81.0  milleryvonne@fowler-york.com  15/10/2016  123018.000000     SupportE  
6  66.0           pperkinshotmail.com  17/09/2020   77311.000000           HR  
7  38.0      josephrobertsongmail.com  2021-01-12            NaN      Support  
8  71.0          andrew74@hotmail.com  2022-02-12   83354.000000        Sales  
9  44.0      denisemaynard@jordan.com  2022-12-08   37701.725993           HR

In [5]:
#last few rows of the data
df.tail(10)

Unnamed: 0                                    ID  \
10990        3317  10c26512-ce3a-4387-b71a-47768503562e   
10991        9603  612011e6-101d-4ee2-bbca-16a5056d8217   
10992        7429  749e71e0-f852-4749-bae3-b59b9c5a9e77   
10993        9510  796f8c0b-15a4-475c-bb02-ac892df0c746   
10994        3978  6714ff24-3600-4c6e-8eed-5b4224bd9e44   
10995        6523  07c223be-03e6-4f70-a2b5-86df778cc61a   
10996        9785  da8a6bbc-5026-4630-848d-f64e80dac56c   
10997        7826  ed19c966-d6d8-4047-b410-b6e595a39340   
10998        7648  783b36b4-d09f-46c9-8a52-7ff96b80863e   
10999        7107  fc25a38a-5747-46eb-b6d3-7173f8255809   

                        Name   Age                      Email   Join Date  \
10990      Kimberly Cummings  54.0            rmccoy@hall.com  2020-06-11   
10991  Jeffrey Elliottnation  72.0        jpatterson@wise.org  2021-08-05   
10992                    NaN  31.0    kingcharles@hotmail.com  2020-11-28   
10993    Shelley Hopkinslive  79.0             Michael Walker  25/12/1992   
10994            Karen Wells  29.0    cheryllewis@hotmail.com         NaN   
10995                    NaN   NaN                        NaN         NaN   
10996       Steven Armstrong  38.0           molly89gmail.com  2021-06-24   
10997     Stephanie Riossell   NaN  robert96@pollard-frye.com  15/08/2006   
10998         Bonnie Benitez  37.0         roypark@warren.net  2020-10-09   
10999         Caroline Ochoa  53.0          cdavis@hodges.com  2023-08-10   

              Salary   Department  
10990   60950.500220    Marketing  
10991   50510.777098      Support  
10992   88530.000000    Marketing  
10993   47809.000000           HR  
10994  147555.000000  Engineering  
10995            NaN          NaN  
10996            NaN        Sales  
10997  122609.594149           HR  
10998  147322.005171      Support  
10999  149224.000000      Support

In [6]:
#getting number of rows & columns
print (df.shape)

(11000, 8)


In [7]:
#identifying types of data in each column
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11000 entries, 0 to 10999
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  11000 non-null  int64  
 1   ID          11000 non-null  object 
 2   Name        8667 non-null   object 
 3   Age         9253 non-null   float64
 4   Email       9731 non-null   object 
 5   Join Date   8808 non-null   object 
 6   Salary      8761 non-null   float64
 7   Department  8745 non-null   object 
dtypes: float64(2), int64(1), object(5)
memory usage: 687.6+ KB


In [8]:
#converting to a pandas datframe
df = pd.DataFrame(data)

In [9]:
#checking type 
print(type(df))
#this is done because sometimes it may be stored as series class

<class 'pandas.core.frame.DataFrame'>


In [10]:
# getting the list of columns
print (df.columns)

Index(['Unnamed: 0', 'ID', 'Name', 'Age', 'Email', 'Join Date', 'Salary',
       'Department'],
      dtype='object')


In [11]:
#removing unrelevant columns
df.drop(columns=['Unnamed: 0'], inplace=True)
#here this column is removed because it is only the index and is not necessary

In [12]:
# Checking for missing values in each column
missing = data.isnull().sum()
# Checking for duplicates
dup_rows = data.duplicated().sum()
# Display summary of missing values and duplicates
missing, dup_rows

(Unnamed: 0       0
 ID               0
 Name          2333
 Age           1747
 Email         1269
 Join Date     2192
 Salary        2239
 Department    2255
 dtype: int64,
 291)

In [13]:
#percentage of missing values
missing_percentage = data.isnull().mean() * 100
print (missing_percentage)

Unnamed: 0     0.000000
ID             0.000000
Name          21.209091
Age           15.881818
Email         11.536364
Join Date     19.927273
Salary        20.354545
Department    20.500000
dtype: float64


In [14]:
#removing all rows with more than 50% of nan values
threshold = len(df.columns) * 0.5
df = df.dropna(thresh=threshold)
#here all the rows with more than 50% nan values cannot be considered as we assume that these rows maybe incomplete and hence imputing them can make the data too unreal

In [15]:
#get size of the dataframe after removing
print (df.shape)

(9725, 7)


In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9725 entries, 0 to 10999
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   ID          9725 non-null   object 
 1   Name        8667 non-null   object 
 2   Age         9249 non-null   float64
 3   Email       9725 non-null   object 
 4   Join Date   8808 non-null   object 
 5   Salary      8760 non-null   float64
 6   Department  8744 non-null   object 
dtypes: float64(2), object(5)
memory usage: 607.8+ KB


In [17]:
# Checking for missing values in each column
missing = df.isnull().sum()
# Checking for duplicates
dup_rows = df.duplicated().sum()
# Display summary of missing values and duplicates
missing, dup_rows

(ID               0
 Name          1058
 Age            476
 Email            0
 Join Date      917
 Salary         965
 Department     981
 dtype: int64,
 114)

In [18]:
#getting percentage of missing value
missing_percentage = df.isnull().mean() * 100
print (missing_percentage)

ID             0.000000
Name          10.879177
Age            4.894602
Email          0.000000
Join Date      9.429306
Salary         9.922879
Department    10.087404
dtype: float64


Here in the above data initially after importing the data there are 8 columns namely Unnamed: 0, ID, Name, Age, Email,	Join Date,	Salary & Department. The data consisted of 11000 rows. ID column and Unnamed: 0 column had zero null values, all other rows consisted of null values. Unnamed: 0 column has only indexes in it and is not required. ID column does not have a specific format and hence may have some relevance since the column is completely filled.After finding the missing values the following are the observations:
##### Unnamed: 0  - 0
##### ID -  0
##### Name  -  2333
##### Age - 1747
##### Email -  1269
##### Join Date - 2192
##### Salary  - 2239
##### Department  -  2255
The above are the corresponding missing values. The percentage of missing value include:
Unnamed: 0 - 0.00%
##### ID - 0.00%
##### Name - 21.20%
##### Age - 15.88%
##### Email - 11.53%
##### Join Date - 19.92%
##### Salary - 20.35%
##### Department - 20.50%

Since there are alot of missing values, rows with more than 50% of null values are removed and thenumber of rows is reduced to 9725 rows. The column Unnamed: 0 is removed as it is not required. After removing the null rows, the missing values of each columns and the percentage missing values is given below:
(ID         0
##### Name       1058
##### Age        476
##### Email      0
##### Join Date  917
##### Salary     965
##### Department 981
 
##### ID         0.00%
##### Name        10.87%
##### Age         4.89%
##### Email       0.00%
##### Join Date   9.42%
##### Salary      9.92%
##### Department  10.08%

Now the data is preped for cleaning.

## Cleaning Data

In [19]:
#checking datatypes of entries
for col in df.columns:
    print(f"{col}: {df[col].apply(type).value_counts()}")

ID: ID
<class 'str'>    9725
Name: count, dtype: int64
Name: Name
<class 'str'>      8667
<class 'float'>    1058
Name: count, dtype: int64
Age: Age
<class 'float'>    9725
Name: count, dtype: int64
Email: Email
<class 'str'>    9725
Name: count, dtype: int64
Join Date: Join Date
<class 'str'>      8808
<class 'float'>     917
Name: count, dtype: int64
Salary: Salary
<class 'float'>    9725
Name: count, dtype: int64
Department: Department
<class 'str'>      8744
<class 'float'>     981
Name: count, dtype: int64


In [20]:
df_cleaned= df.dropna()

In [21]:
#Changing Age column to numeric
df['Age'] = pd.to_numeric(df['Age'], errors='coerce')

In [22]:
#Changing Salary column to numeric
df['Salary'] = pd.to_numeric(df['Salary'], errors='coerce')

In [23]:
# Filling missing values in the 'Age' column with the median age
df['Age'].fillna(df['Age'].median(), inplace=True)
# Here median age is used because it has the middle value of age from the sorted ages, assuming that most people may have the middle most age 

In [24]:
# Filling missing values in the 'Salary' column with the mean salary
df['Salary'].fillna(df['Salary'].mean(), inplace=True)
#mean salary is taken as it will be more or less near to the actual salary.

In [25]:
#rounding to integer value
df_cleaned['Salary'] = np.floor(df['Salary']).astype(int)
#assuming that salary is always a whole number

C:\Users\soft\AppData\Local\Temp\ipykernel_1320\717267621.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_cleaned['Salary'] = np.floor(df['Salary']).astype(int)


In [26]:
#this is the pattern of the valid emails
email_pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'

In [27]:
# Apply regex pattern to find valid emails
df['valid_email'] = df['Email'].apply(lambda x: re.match(email_pattern, str(x)) is not None)
df.head()
#here a new column valid_email is added which has boolean value true and false based on valid email or not

ID                     Name   Age  \
0  1e407ff9-6255-489d-a0de-34135d4f74bd            Hunter Thomas  25.0   
1  379f55b8-87d5-4739-a146-7400b78c24d1             Jeremy Irwin  90.0   
2  18261368-dfa1-47f0-afc6-bddf45926b07  Jennifer Hammondquickly  66.0   
3  ae7cf7cf-17cf-4c8b-9c44-4f61a9a238e5          Sydney Taylorso  39.0   
4  14ed3e6a-e0f5-4bbe-8d93-8665267f5c90                Julia Lee  71.0   

                     Email   Join Date    Salary   Department  valid_email  
0       xlopez@hotmail.com         NaN   88552.0        Sales         True  
1          Jillian Jenkins  2022-07-07  139227.0          NaN        False  
2          jscottgreen.biz  2023-11-21   65550.0  Engineering        False  
3       luke56gonzalez.com  2021-11-05  139932.0     SupportJ        False  
4  figueroakayla@yahoo.com         NaN  143456.0    Marketing         True

In [28]:
#getting all the rows with true valid email ID's
df_cleaned = df[df['valid_email'] == True]

In [29]:
df_cleaned.head()

ID           Name   Age  \
0  1e407ff9-6255-489d-a0de-34135d4f74bd  Hunter Thomas  25.0   
4  14ed3e6a-e0f5-4bbe-8d93-8665267f5c90      Julia Lee  71.0   
5  f92ea13f-24d3-4792-91e4-5b0e07d90c2c    Lisa Bailey  81.0   
8  12cc2f16-d641-402e-b22b-3332bafc4f1d            NaN  71.0   
9  d06f507d-62e5-4bd6-a252-4e60e109d324     Jay Fisher  44.0   

                          Email   Join Date         Salary Department  \
0            xlopez@hotmail.com         NaN   88552.000000      Sales   
4       figueroakayla@yahoo.com         NaN  143456.000000  Marketing   
5  milleryvonne@fowler-york.com  15/10/2016  123018.000000   SupportE   
8          andrew74@hotmail.com  2022-02-12   83354.000000      Sales   
9      denisemaynard@jordan.com  2022-12-08   37701.725993         HR   

   valid_email  
0         True  
4         True  
5         True  
8         True  
9         True

In [30]:
print (df_cleaned.shape)

(7297, 8)


In [31]:
# removing the extra created column
df_cleaned = df_cleaned.drop(columns=['valid_email'])

In [32]:
print (df_cleaned.shape)

(7297, 7)


In [33]:
#to go through column names
column_values = df['Name'].values
column_values = df['Name'].tolist()
print(column_values)

['Hunter Thomas', 'Jeremy Irwin', 'Jennifer Hammondquickly', 'Sydney Taylorso', 'Julia Lee', 'Lisa Bailey', 'Patricia Brown DVM', 'Derek Murphy', nan, 'Jay Fisher', 'Craig Conway', 'Raymond Wilson', 'Amber Marsh', 'Patrick Mooney', 'Nicole Cardenas', nan, 'Mark Wilkins', nan, 'Eric Cole', 'Danny Martinred', 'Curtis Parkerlot', 'Laura Davis', 'Sharon Mcclure', 'Lori Hernandez', 'Christina Norriscompare', 'Amber Garcia', 'Trevor Smith', 'Mark Smith', 'John Weeks', 'Jamie Valdez', nan, 'Ryan DanielsTV', 'Veronica Schmidt', 'Richard Brooks', 'Cesar Morrison', 'Sandra Young', 'Dawn Tucker', 'Kayla Carpenter', 'Laura Douglas', 'Evan Martinezimportant', 'William Avila', 'Kathleen Pitts', 'Alice Lewis', 'Richard Collins', 'John Bailey', 'Joshua Harmonview', 'Angela Waller', 'Joseph Calderon', 'Jason Stephens', 'Jessica Evans', 'Kimberly Page', nan, 'Mrs. Sheri Duffyparty', nan, 'Sydney Robinson', 'Carmen Green', 'John Hurst', 'Dennis Martindoor', 'Robert Peterson', 'Allison Mcknighttrip', 'Mig

In [34]:
print(type(df_cleaned))
#if the type is not 'pandas.core.frame.DataFrame' then use the code: df_cleaned = df_cleaned.to_frame()

<class 'pandas.core.frame.DataFrame'>


In [35]:
# renaming the column since whitespace in column name may cause error 
df.rename(columns={'Join Date': 'Join_Date'}, inplace=True)

In [36]:
df_cleaned.rename(columns={'Join Date': 'Join_Date'}, inplace=True)

In [37]:
df_cleaned.head()

ID           Name   Age  \
0  1e407ff9-6255-489d-a0de-34135d4f74bd  Hunter Thomas  25.0   
4  14ed3e6a-e0f5-4bbe-8d93-8665267f5c90      Julia Lee  71.0   
5  f92ea13f-24d3-4792-91e4-5b0e07d90c2c    Lisa Bailey  81.0   
8  12cc2f16-d641-402e-b22b-3332bafc4f1d            NaN  71.0   
9  d06f507d-62e5-4bd6-a252-4e60e109d324     Jay Fisher  44.0   

                          Email   Join_Date         Salary Department  
0            xlopez@hotmail.com         NaN   88552.000000      Sales  
4       figueroakayla@yahoo.com         NaN  143456.000000  Marketing  
5  milleryvonne@fowler-york.com  15/10/2016  123018.000000   SupportE  
8          andrew74@hotmail.com  2022-02-12   83354.000000      Sales  
9      denisemaynard@jordan.com  2022-12-08   37701.725993         HR

In [38]:
# Convert the 'Join Date' column to datetime
df['Join_Date'] = pd.to_datetime(df['Join_Date'], errors='coerce')

# Format the dates to YYYY-MM-DD
df['Join_Date'] = df['Join_Date'].dt.strftime('%Y-%m-%d')


In [39]:
#Fill nan value with next value
df_cleaned['Join_Date'] = df['Join_Date'].fillna(method='bfill')
#Here a backward fill is applied, assuming that the joining date is clustered and hence details of people joined together are closeby

In [40]:
df_cleaned.head()

ID           Name   Age  \
0  1e407ff9-6255-489d-a0de-34135d4f74bd  Hunter Thomas  25.0   
4  14ed3e6a-e0f5-4bbe-8d93-8665267f5c90      Julia Lee  71.0   
5  f92ea13f-24d3-4792-91e4-5b0e07d90c2c    Lisa Bailey  81.0   
8  12cc2f16-d641-402e-b22b-3332bafc4f1d            NaN  71.0   
9  d06f507d-62e5-4bd6-a252-4e60e109d324     Jay Fisher  44.0   

                          Email   Join_Date         Salary Department  
0            xlopez@hotmail.com  2022-07-07   88552.000000      Sales  
4       figueroakayla@yahoo.com  2021-01-12  143456.000000  Marketing  
5  milleryvonne@fowler-york.com  2021-01-12  123018.000000   SupportE  
8          andrew74@hotmail.com  2022-02-12   83354.000000      Sales  
9      denisemaynard@jordan.com  2022-12-08   37701.725993         HR

In [41]:
#Using unique() to get all unique values in department column
distinct_entries = df['Department'].unique()
print(distinct_entries)

['Sales' nan 'Engineering' 'SupportJ' 'Marketing' 'SupportE' 'HR'
 'Support' 'HRC' 'SalesA' 'SalesX' 'Supportu' 'Engineeringi' 'MarketingR'
 'Salesw' 'SalesK' 'MarketingU' 'Salesi' 'HRa' 'SalesG' 'Supporty'
 'SalesN' 'Marketinge' 'Marketingv' 'HRv' 'SalesM' 'EngineeringL'
 'MarketingQ' 'HRs' 'SalesW' 'Supportj' 'Salesd' 'Marketingu'
 'Engineeringj' 'SalesU' 'HRO' 'MarketingH' 'Salesj' 'Saleso' 'SalesZ'
 'Engineeringe' 'MarketingF' 'Marketingr' 'Engineeringm' 'HRW' 'HRt' 'HRz'
 'Engineeringp' 'HRF' 'EngineeringS' 'SupportW' 'EngineeringT' 'SalesC'
 'SalesB' 'SupportC' 'SalesY' 'Marketingn' 'EngineeringK' 'Marketingz'
 'HRo' 'SupportB' 'HRi' 'Salest' 'Supportg' 'Supporta' 'HRn' 'SupportQ'
 'Salesr' 'EngineeringG' 'Supportm' 'Supporto' 'EngineeringH' 'Marketingw'
 'EngineeringM' 'Supportz' 'HRr' 'Salesy' 'MarketingD' 'HRE' 'Marketingm'
 'HRw' 'Marketingo' 'Supportk' 'MarketingN' 'SupportS' 'EngineeringD'
 'Marketingp' 'Salesl' 'MarketingI' 'MarketingL' 'SupportT' 'Engineeringn'
 'Marketin

In [42]:
# made all entries to string type
df['Department'] = df['Department'].astype(str)

In [43]:
print(df.dtypes)

ID              object
Name            object
Age            float64
Email           object
Join_Date       object
Salary         float64
Department      object
valid_email       bool
dtype: object


In [44]:
# Standardize department names using str.contains (wildcard-like matching)
df_cleaned['Department'] = df['Department'].apply(lambda x: 
    'HR' if pd.Series(x).str.contains('H', case=False).any() else
    'Engineering' if pd.Series(x).str.contains('Engin', case=False).any() else
    'Marketing' if pd.Series(x).str.contains('Mark', case=False).any() else
    'Sales' if pd.Series(x).str.contains('Sal', case=False).any() else
    '' if pd.Series(x).str.contains('nan', case=False).any() else
    'Support' if pd.Series(x).str.contains('Sup', case=False).any() else x
)
# the names of department starts  properly and noise is towards the tail

In [45]:
# Identifying unique department entries (after correction)
print("After correction:", df_cleaned['Department'].unique())
#nan value is denoted with white space

After correction: ['Sales' 'Marketing' 'Support' 'HR' 'Engineering' '']


In [46]:
#filling the empty cells with the most common (mode) value
mode_value = df_cleaned['Department'].mode()[0]  # Get the mode of the column
df_cleaned['Department'].replace('', mode_value, inplace=True)

In [47]:
# Identifying unique department entries (after correction)
print("After correction:", df_cleaned['Department'].unique())

After correction: ['Sales' 'Marketing' 'Support' 'HR' 'Engineering']


In [48]:
# Checking for missing values in each column
miss = df_cleaned.isnull().sum()
# Checking for duplicates
duprows = df_cleaned.duplicated().sum()
# Display summary of missing values and duplicates
miss, duprows

(ID              0
 Name          792
 Age             0
 Email           0
 Join_Date       0
 Salary          0
 Department      0
 dtype: int64,
 126)

In [49]:
# Extract the 'Name' column to a new DataFrame
dfn = df_cleaned[['Name']].copy()

# Clean the 'Name' column
# Remove leading/trailing spaces and fill NaN with 'Unknown'
dfn['Name'] = dfn['Name'].str.strip()  # Remove leading and trailing spaces
dfn['Name'].fillna('Unknown', inplace=True)  # Replace NaN with 'Unknown'
dfn['Name'].replace('', 'Unknown', inplace=True)  # Replace empty strings with 'Unknown'


In [50]:
#checking name column
dfn.head()

Name
0  Hunter Thomas
4      Julia Lee
5    Lisa Bailey
8        Unknown
9     Jay Fisher

In [51]:
#checking for missing values if any
print(dfn.isnull().sum())

Name    0
dtype: int64


In [52]:
# Replace the cleaned 'Name' column back into the original DataFrame
df_cleaned['Name'] = dfn['Name']

In [53]:
#checking if there are any more missing values
print(df_cleaned.isnull().sum())

ID            0
Name          0
Age           0
Email         0
Join_Date     0
Salary        0
Department    0
dtype: int64


In [54]:
#cleaned datafarme
df_cleaned.head(20)


ID                     Name   Age  \
0   1e407ff9-6255-489d-a0de-34135d4f74bd            Hunter Thomas  25.0   
4   14ed3e6a-e0f5-4bbe-8d93-8665267f5c90                Julia Lee  71.0   
5   f92ea13f-24d3-4792-91e4-5b0e07d90c2c              Lisa Bailey  81.0   
8   12cc2f16-d641-402e-b22b-3332bafc4f1d                  Unknown  71.0   
9   d06f507d-62e5-4bd6-a252-4e60e109d324               Jay Fisher  44.0   
10  d5fe9321-49cf-4f14-ac96-7aea1204247b             Craig Conway  88.0   
12  da63d159-e9b0-4b92-b0fe-4e87df3aebc1              Amber Marsh  23.0   
13  5ca27087-5c90-4d42-94d0-7e9ee3f422b9           Patrick Mooney  57.0   
14  9b83a5e5-9516-493f-86f8-d6c445edb501          Nicole Cardenas  60.0   
15  0074dbd5-10d0-463f-9b10-d8473bea04ba                  Unknown  22.0   
16  241e20d7-cc36-495e-86c2-2928bbaa93d8             Mark Wilkins  54.0   
21  7aa89cdd-e328-4425-9556-b0b3da9e4f2a          Danny Martinred  64.0   
22  88e92fab-a06c-4c53-a59a-699bf8a583b5         Curtis Parkerlot  36.0   
23  13b6428a-ba54-4531-87a1-72c3c3ccf765              Laura Davis  60.0   
25  58d4c7cd-aa32-4429-a96e-e881c676852c           Sharon Mcclure  25.0   
26  ea358a83-5e3b-41dd-b00b-9cbb2bb85b45           Lori Hernandez  83.0   
27  2f99edc1-c0e7-4963-8b33-09465a2ac68f  Christina Norriscompare  84.0   
29  31f74cb1-8ed7-464a-9014-5b833cf9e5db             Trevor Smith  30.0   
34  ace748f5-03e6-4187-8a02-c81906eb9667                  Unknown  18.0   
35  6f5313d3-eeb9-4acb-bd8a-47797d747c5a           Ryan DanielsTV  35.0   

                              Email   Join_Date         Salary   Department  
0                xlopez@hotmail.com  2022-07-07   88552.000000        Sales  
4           figueroakayla@yahoo.com  2021-01-12  143456.000000    Marketing  
5      milleryvonne@fowler-york.com  2021-01-12  123018.000000      Support  
8              andrew74@hotmail.com  2022-02-12   83354.000000        Sales  
9          denisemaynard@jordan.com  2022-12-08   37701.725993           HR  
10       gerald47@woodard-smith.com  2021-08-26   81214.525972  Engineering  
12               robert43@gmail.com  2023-09-08  112970.000000  Engineering  
13        bowersjasmine@hotmail.com  2023-06-24   41511.139646        Sales  
14        griffinlinda@harrison.com  2023-03-01   51625.814403    Marketing  
15  ericjohnson@castillo-mendez.net  2023-03-01   89883.332796           HR  
16          longjennifer@parker.biz  2022-01-19   85794.000000        Sales  
21              amyclements@lee.biz  2022-08-30   89883.332796           HR  
22       daniellewilliams@gmail.com  2023-12-08   89883.332796           HR  
23  garciaralph@hayes-hernandez.biz  2024-06-09   61209.000000  Engineering  
25     mcmahonchristina@hotmail.com  2024-06-09   89208.000000    Marketing  
26            elizabeth22@yahoo.com  2021-05-29   52075.000000  Engineering  
27              haleygary@yahoo.com  2021-07-17   92668.000000  Engineering  
29           robertflowers@diaz.biz  2022-06-13  133390.000000    Marketing  
34      jeremyarmstrong@hotmail.com  2023-01-14   35900.000000  Engineering  
35                  hward@yahoo.com  2023-01-14  108317.000000        Sales

Checking for any outliers in salary and age column

In [55]:
# Get the minimum salary
minsal = df_cleaned['Salary'].min()
#Get the maximum salary
maxsal = df_cleaned['Salary'].max()
# Display the results
print("Minimum Salary:", minsal)
print("Maximum Salary:", maxsal)
#assuming that salary range is proper here

Minimum Salary: 26233.921418582304
Maximum Salary: 176156.20674749595


In [56]:
#rounding to integer value
df_cleaned['Salary'] = np.floor(df['Salary']).astype(int)

In [57]:
# Get the minimum salary
minage = df_cleaned['Age'].min()
# Get the maximum salary
maxage = df_cleaned['Age'].max()
# Display the results
print("Minimum Age:", minage)
print("Maximum Age:", maxage)

Minimum Age: 18.0
Maximum Age: 90.0


In [58]:
#we assume people retire after 75 years
df_cleaned = df_cleaned[df_cleaned['Age'] <= 75]

In [59]:
df_cleaned.head(20)

ID              Name   Age  \
0   1e407ff9-6255-489d-a0de-34135d4f74bd     Hunter Thomas  25.0   
4   14ed3e6a-e0f5-4bbe-8d93-8665267f5c90         Julia Lee  71.0   
8   12cc2f16-d641-402e-b22b-3332bafc4f1d           Unknown  71.0   
9   d06f507d-62e5-4bd6-a252-4e60e109d324        Jay Fisher  44.0   
12  da63d159-e9b0-4b92-b0fe-4e87df3aebc1       Amber Marsh  23.0   
13  5ca27087-5c90-4d42-94d0-7e9ee3f422b9    Patrick Mooney  57.0   
14  9b83a5e5-9516-493f-86f8-d6c445edb501   Nicole Cardenas  60.0   
15  0074dbd5-10d0-463f-9b10-d8473bea04ba           Unknown  22.0   
16  241e20d7-cc36-495e-86c2-2928bbaa93d8      Mark Wilkins  54.0   
21  7aa89cdd-e328-4425-9556-b0b3da9e4f2a   Danny Martinred  64.0   
22  88e92fab-a06c-4c53-a59a-699bf8a583b5  Curtis Parkerlot  36.0   
23  13b6428a-ba54-4531-87a1-72c3c3ccf765       Laura Davis  60.0   
25  58d4c7cd-aa32-4429-a96e-e881c676852c    Sharon Mcclure  25.0   
29  31f74cb1-8ed7-464a-9014-5b833cf9e5db      Trevor Smith  30.0   
34  ace748f5-03e6-4187-8a02-c81906eb9667           Unknown  18.0   
35  6f5313d3-eeb9-4acb-bd8a-47797d747c5a    Ryan DanielsTV  35.0   
36  3fc48575-3616-4fa9-a554-610f92c8c681  Veronica Schmidt  40.0   
37  de98cc3c-0802-4824-929f-73eee22aa40b    Richard Brooks  56.0   
38  0bb8f303-5515-429e-a05a-7c9ee675a50b    Cesar Morrison  72.0   
42  3a6b1096-f7a7-4901-a31a-9db676895801   Kayla Carpenter  61.0   

                              Email   Join_Date  Salary   Department  
0                xlopez@hotmail.com  2022-07-07   88552        Sales  
4           figueroakayla@yahoo.com  2021-01-12  143456    Marketing  
8              andrew74@hotmail.com  2022-02-12   83354        Sales  
9          denisemaynard@jordan.com  2022-12-08   37701           HR  
12               robert43@gmail.com  2023-09-08  112970  Engineering  
13        bowersjasmine@hotmail.com  2023-06-24   41511        Sales  
14        griffinlinda@harrison.com  2023-03-01   51625    Marketing  
15  ericjohnson@castillo-mendez.net  2023-03-01   89883           HR  
16          longjennifer@parker.biz  2022-01-19   85794        Sales  
21              amyclements@lee.biz  2022-08-30   89883           HR  
22       daniellewilliams@gmail.com  2023-12-08   89883           HR  
23  garciaralph@hayes-hernandez.biz  2024-06-09   61209  Engineering  
25     mcmahonchristina@hotmail.com  2024-06-09   89208    Marketing  
29           robertflowers@diaz.biz  2022-06-13  133390    Marketing  
34      jeremyarmstrong@hotmail.com  2023-01-14   35900  Engineering  
35                  hward@yahoo.com  2023-01-14  108317        Sales  
36               steven75@yahoo.com  2020-03-13   65482  Engineering  
37             lukewalker@gmail.com  2020-06-14  116303        Sales  
38                 lisa58@gmail.com  2020-06-14  162857           HR  
42                dcooper@adams.net  2021-06-08   82857  Engineering

In [60]:
# Finding the number of duplicate rows
num_duplicates = df_cleaned.duplicated().sum()
print (num_duplicates)

105


In [61]:
# Removing duplicate rows
dff = df_cleaned.drop_duplicates()

In [62]:
print (dff.shape)

(5749, 7)


In [63]:
dff.head(20)

ID              Name   Age  \
0   1e407ff9-6255-489d-a0de-34135d4f74bd     Hunter Thomas  25.0   
4   14ed3e6a-e0f5-4bbe-8d93-8665267f5c90         Julia Lee  71.0   
8   12cc2f16-d641-402e-b22b-3332bafc4f1d           Unknown  71.0   
9   d06f507d-62e5-4bd6-a252-4e60e109d324        Jay Fisher  44.0   
12  da63d159-e9b0-4b92-b0fe-4e87df3aebc1       Amber Marsh  23.0   
13  5ca27087-5c90-4d42-94d0-7e9ee3f422b9    Patrick Mooney  57.0   
14  9b83a5e5-9516-493f-86f8-d6c445edb501   Nicole Cardenas  60.0   
15  0074dbd5-10d0-463f-9b10-d8473bea04ba           Unknown  22.0   
16  241e20d7-cc36-495e-86c2-2928bbaa93d8      Mark Wilkins  54.0   
21  7aa89cdd-e328-4425-9556-b0b3da9e4f2a   Danny Martinred  64.0   
22  88e92fab-a06c-4c53-a59a-699bf8a583b5  Curtis Parkerlot  36.0   
23  13b6428a-ba54-4531-87a1-72c3c3ccf765       Laura Davis  60.0   
25  58d4c7cd-aa32-4429-a96e-e881c676852c    Sharon Mcclure  25.0   
29  31f74cb1-8ed7-464a-9014-5b833cf9e5db      Trevor Smith  30.0   
34  ace748f5-03e6-4187-8a02-c81906eb9667           Unknown  18.0   
35  6f5313d3-eeb9-4acb-bd8a-47797d747c5a    Ryan DanielsTV  35.0   
36  3fc48575-3616-4fa9-a554-610f92c8c681  Veronica Schmidt  40.0   
37  de98cc3c-0802-4824-929f-73eee22aa40b    Richard Brooks  56.0   
38  0bb8f303-5515-429e-a05a-7c9ee675a50b    Cesar Morrison  72.0   
42  3a6b1096-f7a7-4901-a31a-9db676895801   Kayla Carpenter  61.0   

                              Email   Join_Date  Salary   Department  
0                xlopez@hotmail.com  2022-07-07   88552        Sales  
4           figueroakayla@yahoo.com  2021-01-12  143456    Marketing  
8              andrew74@hotmail.com  2022-02-12   83354        Sales  
9          denisemaynard@jordan.com  2022-12-08   37701           HR  
12               robert43@gmail.com  2023-09-08  112970  Engineering  
13        bowersjasmine@hotmail.com  2023-06-24   41511        Sales  
14        griffinlinda@harrison.com  2023-03-01   51625    Marketing  
15  ericjohnson@castillo-mendez.net  2023-03-01   89883           HR  
16          longjennifer@parker.biz  2022-01-19   85794        Sales  
21              amyclements@lee.biz  2022-08-30   89883           HR  
22       daniellewilliams@gmail.com  2023-12-08   89883           HR  
23  garciaralph@hayes-hernandez.biz  2024-06-09   61209  Engineering  
25     mcmahonchristina@hotmail.com  2024-06-09   89208    Marketing  
29           robertflowers@diaz.biz  2022-06-13  133390    Marketing  
34      jeremyarmstrong@hotmail.com  2023-01-14   35900  Engineering  
35                  hward@yahoo.com  2023-01-14  108317        Sales  
36               steven75@yahoo.com  2020-03-13   65482  Engineering  
37             lukewalker@gmail.com  2020-06-14  116303        Sales  
38                 lisa58@gmail.com  2020-06-14  162857           HR  
42                dcooper@adams.net  2021-06-08   82857  Engineering

In [64]:
# exporting the cleaned data to csv file 
dff.to_csv('cleaned_dataset.csv', index=False)